subcommands/targets/offline-update.go

```diff
@@ -366,6 +374,64 @@ func downloadTufRepo(factory string, target string, tag string, prod bool, wave
 	return nil
 }
 
+func copyFile(src string, dst string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+
+	dstFile, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer dstFile.Close()
+
+	_, err = dstFile.ReadFrom(srcFile)
+	if err != nil {
+		return err
+	}
+	srcInfo, err := os.Stat(src)
+	if err != nil {
+		return err
+	}
+	return os.Chmod(dst, srcInfo.Mode())
+}
+
+func copyDirRecursive(srcDir string, dstDir string) error {
+	srcInfo, err := os.Stat(srcDir)
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(dstDir, srcInfo.Mode())
+	if err != nil {
+		return err
+	}
+
+	dirEntries, err := os.ReadDir(srcDir)
+	if err != nil {
+		return err
+	}
+	for _, dirEntry := range dirEntries {
+		src := path.Join(srcDir, dirEntry.Name())
+		dst := path.Join(dstDir, dirEntry.Name())
+		if dirEntry.IsDir() {
+			err = copyDirRecursive(src, dst)
+		} else {
+			err = copyFile(src, dst)
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func copyOstree(srcDir string, dstDir string) error {
+	return copyDirRecursive(srcDir, dstDir)
```

Comment:
Some time in the future (with Golang 1.23) we might use `os.CopyFS`.
For now, using `filepath.Walk` seems optimal to me e.g. smth like below:

```
import (
	"os"
	"os/filepath"
)

func copyRecursive(src, dst string) error {
	copyFile = func(src, dst string) error {
		// Basically, your copyFile function less the fs.Stat nad os.Chmod.
		srcFile, err := os.Open(src)
		if err != nil {
			return nil
		}
		defer srcFile.Close()

		dstFile, err := os.Create(dst)
		if err != nil {
			return err
		}
		defer dstFile.Close()

		return dstFile.ReadFrom(srcFile)
	}

	return filepath.Walk(src, func(srcPath string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		dstPath := strings.Join(dst, strings.Replace(srcPath, src, "", 1))
		if info.IsDir() {
			return os.Mkdir(dstPath, info.Mode())
		} else {
			if err = copyFile(srcPath, dstPath); err != nil {
				return err
			}
			return os.Chmod(dstPath, info.Mode())
		} 
	})
}
```

Reply:
A few notes on the above:
1. Using `filepath.Walk` saves some lines of code.
2. IIUC `os.Chmod` needs to be called after `file.Close` to avoid concurrent file writes (which are dangerous). That's why `os.Chmod` is called outside a function with deferred `file.Close` call.
3. IIRC `filepath.Walk` replaces a stack-based recursion with a tree and a loop (more efficient).
4. You call `os.Stat` on each dir entry, hence, leveraging `Walk` (and not `WalkDir`) makes sense.

Reply:
Thanks, that makes sense. I've updated the code following your suggestion (with minor adjustments), and functionality was not affected.